core/modules/flowgen.cc

```diff
@@ -231,6 +235,72 @@ pb_error_t FlowGen::InitFlowPool() {
   return pb_errno(0);
 }
 
+void FlowGen::UpdateDerivedParameters() {
+  /* calculate derived variables */
+  pareto_.inversed_alpha = 1.0 / pareto_.alpha;
+
+  if (duration_ == DURATION_PARETO) {
+    MeasureParetoMean();
+  }
+
+  concurrent_flows_ = flow_rate_ * flow_duration_;
+  if (concurrent_flows_ > 0.0) {
+    flow_pps_ = total_pps_ / concurrent_flows_;
+  }
+
+  flow_pkts_ = flow_pps_ * flow_duration_;
+  if (flow_rate_ > 0.0) {
+    flow_gap_ns_ = 1e9 / flow_rate_;
+  }
+}
+
+pb_cmd_response_t FlowGen::CommandUpdate(const bess::pb::FlowGenArg &arg) {
+  pb_cmd_response_t response;
+  if(arg.template_().length() > 0){
+    LOG(INFO) << "Updating FlowGen template";
+    if (arg.template_().length() > MAX_TEMPLATE_SIZE) {
+        set_cmd_response_error(&response, pb_error(EINVAL, "'template' is too big"));
+        return response;
+    }
+    template_size_ = arg.template_().length();
+
+    memset(templ_, 0, MAX_TEMPLATE_SIZE);
+    memcpy(templ_, arg.template_().c_str(), template_size_);
+  }
+
+  if (!(std::isnan(arg.pps()) || arg.pps() <= 0.0)) {
+    LOG(INFO) << "Updating FlowGen pps" << arg.pps();
+    total_pps_ = arg.pps();
+  }
+
+  if (!(std::isnan(arg.flow_rate()) || arg.flow_rate() <= 0.0)) {
+    LOG(INFO) << "Updating FlowGen flow rate" << arg.flow_rate();
+    flow_rate_ = arg.flow_rate();
+  }
+
+  if (!(std::isnan(arg.flow_duration()) || arg.flow_duration() <= 0.0)) {
+    LOG(INFO) << "Updating FlowGen flow duration" << arg.flow_duration();
+    flow_duration_ = arg.flow_duration();
+  }
+
+  if (arg.arrival() == "uniform") {
+    arrival_ = ARRIVAL_UNIFORM;
+  } else if (arg.arrival() == "exponential") {
+    arrival_ = ARRIVAL_EXPONENTIAL;
+  }
+
+  if (arg.duration() == "uniform") {
+    duration_ = DURATION_UNIFORM;
+  } else if (arg.duration() == "pareto") {
+    duration_ = DURATION_PARETO;
+  }
+
+  UpdateDerivedParameters();
+
+  set_cmd_response_error(&response, pb_errno(0));
```

Comment:
What I mean by "this function" was `CommandUpdate()`. Unlike `Init()`, `CommandUpdate()` is skipping most of parameter error returns, except the packet template. For exmple, if a negative pps given by a buggy controller will be silently ignored.